### ISSUE TITLE
Porting EF6 to EF Core: error "The LINQ expression could not be translated" reported in lots of queries

### ISSUE THREAD

**Issue by phlogan** [dotnet/efcore#29404]
Hello everyone.

Recently, me and my team decided to migrate our whole web system from .NET Framework to .NET 6, and upgrade/migrate all of our frameworks too, this lead us to migrate from **EF6** to **EF Core**.

We are almost done, but a lot of queries are reporting the error "The LINQ expression could not be translated", all of them with basically the same structure as shown below:

`IEnumerable<Guid> tagIds;//this one comes often from the methods parameters, sometimes as List<Guid>`
`var result = Db.Products.Where(e => tagIds.Any(tagId => e.Tags.Any(tg => tg.Id == tagId))).ToList();`
(The error is generated because of the code inside the "Where" clause)

**In short, we need to compare a list to database fields (i.e. filter data by list of IDs)**, similar to the "IN" operator from SQL.

There is any way to do it **WITHOUT "ToList" or "FirstOrDefault"** before the "Where"?

 Thanks!

**Comment by Xor-el** [comment]
@phlogan if I get your question correctly, replacing your existing query with this below should allow EF Core to translate it correctly.

```cs
IEnumerable<Guid> tagIds;//this one comes often from the methods parameters, sometimes as List<Guid>
var result = Db.Products.Where(e => e.Tags.Any(tg => tagIds.Contains(tg.Id))).ToList();
```

**Comment by phlogan** [comment]
> @phlogan if I get your question correctly, replacing your existing query with this below should allow EF Core to translate it correctly.
> 
> ```cs
> IEnumerable<Guid> tagIds;//this one comes often from the methods parameters, sometimes as List<Guid>
> var result = Db.Products.Where(e => e.Tags.Any(tg => tagIds.Contains(tg.Id))).ToList();
> ```

Hello!

Thanks, this works for that case i've quoted!

We are now replacing the code, for now it's working in most cases, except for that specific:
`var allLanguages = ((LanguageEnum[])Enum.GetValues(typeof(LanguageEnum))).ToList()`
`return DbSet`
`                .Include(e => e.Translations)`
`                .Select(e => new Tag`
`                {`
                    `//Checks if the current Tag Translations has all values set in the "allLanguages" list`
`                    HasAllPossibleTranslations = !allLanguages.Any(t => !e.Translations.Select(f => f.Language).Contains(t))`
`                }).FirstOrDefault();`

We need to compare the list to the database's TagTranslations, in that order, and can not compare by quantity (Count), because the Tag can have more than 1 Translation per Language.

But the good news is that we found a workaround that fitted well. First, we've created an extension method:
        `public static bool ContainsAll<T>(this IEnumerable<T> source, IEnumerable<T> target)`
        `{`
            `return source.All(e => target.Contains(e));`
        `}`
And then, replaced the code to:
`HasAllPossibleTranslations = !allLanguages.ContainsAll(e.Translations.Select(f => f.Language))`

And that's it! As I said, for now it's everything working.

I consider this issue closed, once again, thanks a lot for the help!
